Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1380 @@
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Formal grammar 
+
+When talking about parsing, we generally imply we're parsing some language, a formal language in the case of programming languages.
+But what is a language?
+
+Formal language theory offers a very simple answer: a language is a (likely infinite) set of all grammatically correct sentences.
+Now, the word "sentence" is used in a very general sense here – a complete MiniML program would still constitute a sentence.
+
+Well, but how do you describe an infinite set of sentences?
+You do that by specifying how how to build any arbitrary sentence.
+
+This description is called the formal grammar of the language.
+The usual approach uses string substitutions to construct a sentence.
+The process of constructing a sentence itself is called the derivation process, which is represented as a sequence of substitutions called a derivation chain.
+We can refer to either as **derivation**.
+
+We start with a special symbol called the start non-terminal. Then, we find and replace symbols according to our substitution rules. Once we have replaced everything we could have replaced, we stop.
+
+The symbols to be replaced are called **non-terminals**, and the symbols they are ultimately replaced by are called **terminals**. Non-terminals denote things that do not occur in the language itself. Terminals are usually the lexical items of the language – you can think of them being the "words" of the language. They are called so because derivation terminates on them.
+
+The rules for substituting symbols are called **production rules**.
+They are the rules that define the grammar (and, by extension, the language).
+In formal language theory, we usually write them as `alpha -> beta`, where `alpha` is called the production head, and `beta` the production body.
+In each step of derivation, some head is replaced by its corresponding body.
+
+### Parsing with a grammar
+
+Parsing is, in a sense, a process inverse to derivation. Instead of building an arbitrary sentence, we're given a sentence and tasked with recovering the derivation that produced it. 
+
+There are two broad approaches to accomplish that. We could try to run the derivation process forwards, but at each step choose the substitution that would get us closer to the input string. This is top-down parsing. Not all grammars work well with this approach, but it is arguably more straightforward. 
+
+Or we could run the derivation process in reverse by finding a production rule whose body matches some of the input string. We do that until we're back at the starting non-terminal. This is bottom-up parsing. It works well with more grammars (still not all of them, though), and that's the approach Happy adheres to.
+
+In our case, we'll generate a bottom-up parser that will build an **abstract syntax tree** (AST), which is a data structure that describes the exact structure of the language we're parsing.
```

Comment:
s/exact structure of the language we're parsing/exact grammatical structure of the input/ perhaps?